File: client/get_token.py
import requests
import os
import re
import base64

class Token():

    def __init__(self, *env):
        env = env[0] if len(env) > 0 else '.env'
        self.env = self.load_env(env)
        try:
            self.scope = self.env['SCOPE']
            self.url = f"{self.env['OKTAISSUER']}/v1/token"
            self.key = self.env['OKTACLIENTID'];
            self.secret = self.env['OKTASECRET']
        except:
            pass

    def load_env(self, env):
        cwd = f"{os.getcwd()}/{env}"
        if (not os.path.isfile(cwd)):
            cwd = f"{os.getcwd()}/../{env}"
            if (not os.path.isfile(cwd)):
                return None
        file = None
        with open(cwd, 'r') as f:
            file = [line[:-1] for line in f.readlines()]
        file = {re.match('[^=]+', line)[0]: re.search('[^=]+$', line)[0] for line in file}
        return file

    def get_token(self):
        if not self.env:
            return None;
        payload = f"grant_type=client_credentials&scope={self.scope}"
        token = f"{self.key}:{self.secret}"
        token = str(base64.b64encode(bytes(token, encoding='utf-8')), encoding='utf-8')
        headers = {
                "Content-Type": "application/x-www-form-urlencoded",
                "Authorization": f"Basic {token}"
        }

        resp = requests.post(self.url, data=payload, headers=headers, timeout=5)
        if (not(resp.status_code >= 200 and resp.status_code < 300)):
            print(f'Error: {resp.status_code} ({resp.reason})')
            sys.exit(1)
            # raise ConnectionError
        jsonpy = resp.json()
        return f"{jsonpy['token_type']} {jsonpy['access_token']}"
        # return f"{jsonpy['access_token']}"

# token = Token()
# token.get_token()

File: client/get_token.mjs
import fetch from 'node-fetch';
import path from 'path';
import { fileURLToPath } from 'url';
import { readFileSync } from 'node:fs';

export class Token {
  constructor(env = '.env') {
    this.urisuffix = '/v1/token';
    this.loadEnv(env);
  }

  loadEnv(env) {
    const __filename = fileURLToPath(import.meta.url);
    const __dirname = path.dirname(__filename);
    let envPath = __dirname + '/' + env;
    this.readFileRecursive(envPath);
  }

  readFileRecursive(envPath) {
    let env = envPath.match(/[^\/]+$/)[0];
    try {
      let obj = {};
      let data = readFileSync(envPath, 'utf8');
      data
        .split('\n')
        .filter(line => line != '')
        .forEach(line => (obj[line.replace(/=.*$/, '')] = line.replace(/^.*=/, '')));

      for (let [key, value] of Object.entries(obj)) this[key] = value === '0' ? false : value === '1' ? true : value;
    } catch (err) {
      envPath = envPath.replace(/[^\/]+\/([^\/]+)$/, '$1');
      if (envPath != `/${env}`) this.readFileRecursive(envPath);
      else throw new Error(`${env} file not found`);
    }
  }

  async getTokenAsync() {
    let payload = `grant_type=client_credentials&scope=${this.SCOPE}`;
    let token = Buffer.from(this.OKTACLIENTID + ':' + this.OKTASECRET).toString('base64');
    let url = this.OKTAISSUER + this.urisuffix;
    try {
      const options = {
        method: 'post',
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
          Authorization: `Basic ${token}`,
        },
        body: payload,
      };
      let result = await fetch(url, options);
      let json = await result.json();
      return Promise.resolve(json.token_type + ' ' + json.access_token);
    } catch (err) {
      throw new Error('Error: ' + err);
    }
  }
}

//const token = new Token();
//token.getTokenAsync().then(result => console.log(result));
